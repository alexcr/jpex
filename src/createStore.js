import { observable } from 'mobx';
import jsonpath from 'jsonpath';

// This function implements a cheaper JSON.stringify method that
// allows us to create a truncated representation of the given
// object without having to stringify the whole object first,
// saving a lot of time for huge JSON objects.
// It stringifies each property recursively until the resulting
// string reaches a minimum number of characters or the whole
// object is stringified (whichever happens first).
function limitedStringify(json, minChars = Infinity) {
  const props = [];
  let countChars = 2; // start with '{ ' or '[ '

  const isArray = Array.isArray(json);
  const entries = isArray ? Array.from(json.entries()) : Object.entries(json);
  for (let i = 0; i < entries.length && countChars < minChars; i++) {
    const [key, value] = entries[i];
    countChars += isArray ? 0 : key.length + 2; // '${key}: '
    const parsedValue = (
      value !== null && typeof value === 'object' ?
      limitedStringify(value, minChars - countChars) :
      String(value)
    );
    countChars += parsedValue.length;
    props.push(isArray ? parsedValue : `${key}: ${parsedValue}`);
    countChars += 2; // ', ' or ' }' or ' ]'
  }

  return (
    isArray ?
    '[ ' + props.join(', ') + ' ]' :
    '{ ' + props.join(', ') + ' }'
  );
}

// This function converts a JSON object into a flat list of nodes.
// This flat list allows us to render the JSON tree as a virtualized
// list with react-window. Each node contains all information we
// need to make the list look visually like a tree.
// It returns an array with both the nodes list and a path map, which
// is an object that maps each path to its index in the list.
function parseJson(json) {
  const parsedJson = {
    nodeList: [],
    pathMap: {}
  };

  // Use the current time to generate a unique ID for this JSON object.
  // This will be used in conjunction with the node index as the key
  // property when rendering the list. It's safe to use the index as
  // key in this case because we guarantee that the node index never
  // changes for a given node, so we can use it to uniquely identify
  // each node within a JSON object.
  const jsonId = new Date().getTime();

  // This internal function creates a node object with all the information
  // we need and adds it to the node list.
  const addListNode = (key, value, parent, path) => {
    const nodeIndex = parsedJson.nodeList.length;
    const nodeId = `${jsonId}-${nodeIndex}`;
    const depth = path.length - 1;
    const isObject = value !== null && typeof value === 'object';
    const isArray = Array.isArray(value);
    const isScalar = !isArray && !isObject;
    const isArrayItem = parent && parent.isArray;

    const listNode = {
      nodeIndex, nodeId, key, value, depth, parent, isScalar, isArray, isObject, isArrayItem
    };

    parsedJson.nodeList.push(listNode);
    return listNode;
  };

  // This internal function adds a path to the pathMap object. Each
  // path added maps to an array of node indices. This is because the
  // map key is a stringified version of the path, which might not be
  // unique. For example, this JSON object has two paths with the same
  // stringified path a>b: {"a": {"b": 1}, "a>b": 2}
  // When using the map, we compare the actual path array with each
  // matched path array under its stringified path key. For huge JSON
  // objects, this is much cheaper than comparing it with ALL paths.
  const addNodePath = (path, nodeIndex) => {
    const stringifiedPath = path.join('>')
    if (!parsedJson.pathMap[stringifiedPath]) {
      parsedJson.pathMap[stringifiedPath] = [];
    }
    parsedJson.pathMap[stringifiedPath].push({path, nodeIndex});
  };

  // This internal function is the recursive function responsible for
  // traversing the whole JSON tree calling the other functions above.
  const parseNode = (node, path = [], parent = null) => {
    const entries = (
      Array.isArray(node) ?
      Array.from(node.entries()) :
      Object.entries(node)
    );
    return entries.forEach(([key, value]) => {
      const nextPath = [...path, key];
      const listNode = addListNode(key, value, parent, path);
      addNodePath(nextPath, listNode.nodeIndex);

      if (listNode.isArray || listNode.isObject) {
        parseNode(value, nextPath, listNode);
      }
    })
  };

  parseNode({'$': json});
  return parsedJson;
}

export function createStore() {
  return observable.object({
    /* State */

    invalidJson: false,
    invalidQuery: false,
    query: '',
    matchedNodes: new Set(),
    collapsedNodes: new Set(),

    /* Computeds */

    get parsedJson() {
      return this.json ? parseJson(this.json) : {
        nodeList: [],
        pathMap: {}
      }
    },

    get nodeList() {
      return this.parsedJson.nodeList;
    },

    get pathMap() {
      return this.parsedJson.pathMap;
    },

    get visibleNodeList() {
      return this.nodeList.filter(node => (
        !!node.parent && !this.isNodeCollapsed(node.parent)
      ));
    },

    /* Helpers */

    isNodeCollapsed(node) {
      return (
        this.collapsedNodes.has(node.nodeIndex) ||
        (!node.parent ? false : this.isNodeCollapsed(node.parent))
      );
    },

    isNodeMatched(node) {
      return (
        this.matchedNodes.has(node.nodeIndex) ||
        (!node.parent ? false : this.isNodeMatched(node.parent))
      );
    },

    getNodePreview(node) {
      // 100 characters should be enough to fill in the entire
      // width of our tree container.
      return limitedStringify(node.value, 100);
    },

    /* Actions */

    setJSON(json) {
      this.json = json;
      this.invalidJson = json === null;
      this.collapsedNodes.clear();
      this.query = '';
    },

    setQuery(query) {
      this.query = query;
      this.matchedNodes.clear();

      let matchedPaths;
      try {
        matchedPaths = jsonpath.paths(this.json, query);
        this.invalidQuery = false;
      } catch(e) {
        this.invalidQuery = (query !== '');
      }

      if (matchedPaths) {
        matchedPaths.forEach(path => {
          // Find the node index mapped to the given path.
          const mappedPath = this.pathMap[path.join('>')].find(pathCandidate => {
            return pathCandidate.path.every((p, i) => p === path[i])
          });

          // Now add the found node index to the matchedNodes set.
          // We expect all paths found by jsonpath to exist in the pathMap,
          // so we don't need to check if (mappedPath !== undefined) here.
          this.matchedNodes.add(mappedPath.nodeIndex);
        })
      }
    },

    collapseNode(nodeIndex) {
      this.collapsedNodes.add(nodeIndex);
    },

    expandNode(nodeIndex) {
      this.collapsedNodes.delete(nodeIndex);
    },
  });
}
