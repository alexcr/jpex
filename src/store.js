import React from 'react';
import { useLocalStore } from 'mobx-react-lite';
import { createStore } from './createStore';

const storeContext = React.createContext(null);

export const StoreProvider = ({children}) => {
  const store = useLocalStore(createStore);
  return (
    <storeContext.Provider value={store}>
      {children}
    </storeContext.Provider>
  );
}

export const useStore = () => {
  const store = React.useContext(storeContext);
  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }
  return store;
}
