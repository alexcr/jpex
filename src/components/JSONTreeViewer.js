import React from 'react';
import { observer } from 'mobx-react-lite';
import styled from 'styled-components';
import { FixedSizeList as List } from 'react-window';
import { JSONTreeNode } from './JSONTreeNode';
import { useStore } from '../store';

const StyledList = styled(List)`
  text-align: left;
  margin-left: 15px;
  overflow-x: hidden!important;

  &.error {
    background: url(icon-invalid.png) no-repeat center;
    background-position: calc(50% - 7px);
    background-size: 25%;
  }

  .dragging &.error {
    background: none;
  }
`;

export const JSONTreeViewer = observer(() => {
  const store = useStore();
  const nodeList = store.visibleNodeList;
  return (
    <>
      {store.invalidJson && 'Invalid JSON file'}
      <StyledList
        height={store.invalidJson ? 366 : 380}
        itemCount={nodeList.length}
        itemSize={25}
        className={store.invalidJson ? 'error' : ''}
        itemKey={index => nodeList[index].nodeId}
      >
        {
          ({index, style}) => (
            <JSONTreeNode
              index={index}
              style={style}
            />
          )
        }
      </StyledList>
    </>
  );
});
