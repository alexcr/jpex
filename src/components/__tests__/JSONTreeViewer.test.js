import React from 'react';
import { render } from 'enzyme';
import { JSONTreeViewer } from '../JSONTreeViewer';
import { storeContext } from '../../store';
import { createStore } from '../../createStore';
import { testJson } from '../../__tests__/store.test';

describe('JSONTreeViewer', () => {
  it('should render the tree viewer correctly', () => {
    const store = createStore();
    store.setJSON(testJson);
    expect(
      render(
        <storeContext.Provider value={store}>
          <JSONTreeViewer />
        </storeContext.Provider>
      )
    ).toMatchSnapshot();
  });
});
