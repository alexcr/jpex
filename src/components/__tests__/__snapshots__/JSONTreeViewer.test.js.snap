// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONTreeViewer should render the tree viewer correctly 1`] = `
<div
  class="sc-eCApGN kIhJZd"
  style="position:relative;height:380px;overflow:auto;-webkit-overflow-scrolling:touch;will-change:transform;direction:ltr"
>
  <div
    style="height:525px;width:100%"
  >
    <div
      class="sc-gtssRu hFIiXf"
      style="position:absolute;left:0;top:0;height:25px;width:100%"
    >
      <div
        class="sc-bdnylx dnStey"
      >
        [-]
      </div>
      <span
        class="sc-dlnjPT jLMhUk"
      >
        cities: 
      </span>
      <span
        class="sc-hKFyIo jhwTzf"
      />
    </div>
    <div
      class="sc-gtssRu bwPoAN"
      style="position:absolute;left:0;top:25px;height:25px;width:100%"
    >
      <div
        class="sc-bdnylx dnStey"
      >
        {-}
      </div>
      <span
        class="sc-dlnjPT jLMhUk"
      />
      <span
        class="sc-hKFyIo jhwTzf"
      />
    </div>
    <div
      class="sc-gtssRu JttRD"
      style="position:absolute;left:0;top:50px;height:25px;width:100%"
    >
      <span
        class="sc-dlnjPT jLMhUk"
      >
        city: 
      </span>
      <span
        class="sc-hKFyIo jhwTzf"
      >
        SÃ£o Paulo
      </span>
    </div>
    <div
      class="sc-gtssRu JttRD"
      style="position:absolute;left:0;top:75px;height:25px;width:100%"
    >
      <span
        class="sc-dlnjPT jLMhUk"
      >
        country: 
      </span>
      <span
        class="sc-hKFyIo jhwTzf"
      >
        Brazil
      </span>
    </div>
    <div
      class="sc-gtssRu bwPoAN"
      style="position:absolute;left:0;top:100px;height:25px;width:100%"
    >
      <div
        class="sc-bdnylx dnStey"
      >
        {-}
      </div>
      <span
        class="sc-dlnjPT jLMhUk"
      />
      <span
        class="sc-hKFyIo jhwTzf"
      />
    </div>
    <div
      class="sc-gtssRu JttRD"
      style="position:absolute;left:0;top:125px;height:25px;width:100%"
    >
      <span
        class="sc-dlnjPT jLMhUk"
      >
        city: 
      </span>
      <span
        class="sc-hKFyIo jhwTzf"
      >
        London
      </span>
    </div>
    <div
      class="sc-gtssRu JttRD"
      style="position:absolute;left:0;top:150px;height:25px;width:100%"
    >
      <span
        class="sc-dlnjPT jLMhUk"
      >
        country: 
      </span>
      <span
        class="sc-hKFyIo jhwTzf"
      >
        United Kingdom
      </span>
    </div>
    <div
      class="sc-gtssRu bwPoAN"
      style="position:absolute;left:0;top:175px;height:25px;width:100%"
    >
      <div
        class="sc-bdnylx dnStey"
      >
        {-}
      </div>
      <span
        class="sc-dlnjPT jLMhUk"
      />
      <span
        class="sc-hKFyIo jhwTzf"
      />
    </div>
    <div
      class="sc-gtssRu JttRD"
      style="position:absolute;left:0;top:200px;height:25px;width:100%"
    >
      <span
        class="sc-dlnjPT jLMhUk"
      >
        city: 
      </span>
      <span
        class="sc-hKFyIo jhwTzf"
      >
        Krakow
      </span>
    </div>
    <div
      class="sc-gtssRu JttRD"
      style="position:absolute;left:0;top:225px;height:25px;width:100%"
    >
      <span
        class="sc-dlnjPT jLMhUk"
      >
        country: 
      </span>
      <span
        class="sc-hKFyIo jhwTzf"
      >
        Poland
      </span>
    </div>
    <div
      class="sc-gtssRu hFIiXf"
      style="position:absolute;left:0;top:250px;height:25px;width:100%"
    >
      <div
        class="sc-bdnylx dnStey"
      >
        {-}
      </div>
      <span
        class="sc-dlnjPT jLMhUk"
      >
        duplicatePathTest: 
      </span>
      <span
        class="sc-hKFyIo jhwTzf"
      />
    </div>
    <div
      class="sc-gtssRu bwPoAN"
      style="position:absolute;left:0;top:275px;height:25px;width:100%"
    >
      <div
        class="sc-bdnylx dnStey"
      >
        {-}
      </div>
      <span
        class="sc-dlnjPT jLMhUk"
      >
        a: 
      </span>
      <span
        class="sc-hKFyIo jhwTzf"
      />
    </div>
    <div
      class="sc-gtssRu JttRD"
      style="position:absolute;left:0;top:300px;height:25px;width:100%"
    >
      <span
        class="sc-dlnjPT jLMhUk"
      >
        b: 
      </span>
      <span
        class="sc-hKFyIo jhwTzf"
      >
        first path a&gt;b
      </span>
    </div>
    <div
      class="sc-gtssRu bwPoAN"
      style="position:absolute;left:0;top:325px;height:25px;width:100%"
    >
      <span
        class="sc-dlnjPT jLMhUk"
      >
        a&gt;b: 
      </span>
      <span
        class="sc-hKFyIo jhwTzf"
      >
        second path a&gt;b
      </span>
    </div>
    <div
      class="sc-gtssRu hFIiXf"
      style="position:absolute;left:0;top:350px;height:25px;width:100%"
    >
      <div
        class="sc-bdnylx dnStey"
      >
        [-]
      </div>
      <span
        class="sc-dlnjPT jLMhUk"
      >
        typeTest: 
      </span>
      <span
        class="sc-hKFyIo jhwTzf"
      />
    </div>
    <div
      class="sc-gtssRu bwPoAN"
      style="position:absolute;left:0;top:375px;height:25px;width:100%"
    >
      <span
        class="sc-dlnjPT jLMhUk"
      />
      <span
        class="sc-hKFyIo jhwTzf"
      >
        1.2
      </span>
    </div>
    <div
      class="sc-gtssRu bwPoAN"
      style="position:absolute;left:0;top:400px;height:25px;width:100%"
    >
      <span
        class="sc-dlnjPT jLMhUk"
      />
      <span
        class="sc-hKFyIo jhwTzf"
      >
        true
      </span>
    </div>
    <div
      class="sc-gtssRu bwPoAN"
      style="position:absolute;left:0;top:425px;height:25px;width:100%"
    >
      <span
        class="sc-dlnjPT jLMhUk"
      />
      <span
        class="sc-hKFyIo jhwTzf"
      >
        false
      </span>
    </div>
  </div>
</div>
`;
